File: zadanie2/src/main/java/mastercoder/pkw/mandates/MandateMethodDHondta.java
package mastercoder.pkw.mandates;

import java.util.ArrayList;
import java.util.List;

public class MandateMethodDHondta implements IMandateMethod {

    @Override
    public List<Mandate> calculateMandates(List<Vote> votes, int mandatesCount) {
        List<Vote> v = votes;
        List<Mandate> partMandates = initializeMandates(v);
        List<Double> tab = new ArrayList<>();
        for (Vote vote : v) {
            tab.add((double) vote.validVotes);
        }
        int maxInd = 0;
        for (int i = mandatesCount; i > 0; i--) {
            double max = -1;
            for (int j = 0; j < tab.size(); j++) {
                if (max < tab.get(j)) {
                    max = tab.get(j);
                    maxInd = j;
                }
            }

            partMandates.get(maxInd).mandates++;
            tab.set(maxInd, calc(v.get(maxInd).validVotes, partMandates.get(maxInd).mandates));
        }
        return partMandates;
    }

    private List<Mandate> initializeMandates(List<Vote> v) {
        List<Mandate> partMandates = new ArrayList<>(v.size());
        for (Vote vote : v) {
            Mandate mandate = new Mandate(vote.partShortName, vote.partName, 0);
            partMandates.add(mandate);
        }
        return partMandates;
    }

    private double calc(int v, int mandates) {
        return (double)v * 1.0 / (2*mandates + 1.0);
    }
}

File: zadanie2/MasterCoder.PKW.Mandates/MandateMethodDHondta.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MasterCoder.PKW.Mandates
{
    internal class MandateMethodDHondta : IMandateMethod
    {
        public List<Mandate> CalculateMandates(List<Vote> votes, int mandatesCount)
        {
            var v = votes;
            List<Mandate> partMandates = InitializeMandates(v);

            List<double> tab = new List<double>(v.Select(x => (double)x.ValidVotes));

            int maxInd = 0;

            for (int i = mandatesCount; i > 0; i--)
            {
                double max = -1;
                for (int j = 0; j < tab.Count; j++)
                {
                    if (max < tab[j])
                    {
                        max = tab[j];
                        maxInd = j;
                    }
                }

                partMandates[maxInd].Mandates++;
                tab[maxInd] = Calc(v[maxInd].ValidVotes, partMandates[maxInd].Mandates);
            }

            return partMandates;
        }

        private List<Mandate> InitializeMandates(List<Vote> v)
        {
            List<Mandate> partMandates = new List<Mandate>(v.Count);

            foreach(var vote in v)
            {
                var mandate = new Mandate(vote.PartShortName, vote.PartName, 0);
                partMandates.Add(mandate);
            }

            return partMandates;
        }

        private double Calc(int v, int mandates)
        {
            return ((double)v * 1.0 / (mandates + 1.0)); //w wzorze  [v / (mandates + 1)] nie ma 2*
        }
    }
}
